### PR TITLE
Bump to mono:2020-02

### PR DESCRIPTION
**Pull request by directhex**

### CODE REVIEW

**Comment by radical** [comment]
For the `mac build`:

`Project "$(OutputPath)/../../../xbuild/Xamarin/Android/Xamarin.Android.CSharp.targets" was not imported by "/Users/builder/azdo/_work/3/s/xamarin-android/src/Xamarin.Android.NUnitLite/Xamarin.Android.NUnitLite.csproj" at (355,3), due to false condition; (Exists('$(OutputPath)\..\..\..\xbuild\Xamarin\Android\Xamarin.Android.CSharp.targets')) was evaluated as (Exists('/Users/builder/azdo/_work/3/s/xamarin-android/bin/Release/lib/xamarin.android/xbuild-frameworks/MonoAndroid/v1.0\..\..\..\xbuild\Xamarin\Android\Xamarin.Android.CSharp.targets')).`

- Based on the logs, this seems to be using mono `6.10.0`. Shouldn't this be using `6.12.0`, because of the change in this PR?
- Which exact version of mono is this using?
- Would this path exist `/Users/builder/azdo/_work/3/s/xamarin-android/bin/Release/lib/xamarin.android/xbuild-frameworks/MonoAndroid/v1.0\..\..\..\xbuild\Xamarin\Android\Xamarin.Android.CSharp.targets'`?
- I tried to a test project with `6.10.0.65` to see if it might have regressed, and caused an issue with the conversion of path separators, but it worked fine.

**Comment by jonathanpeppers** [comment]
@radical we normally bump the system Mono here, too:

https://github.com/xamarin/xamarin-android/blob/b954e341fda89d2957e9b0878a35b114f81a73da/build-tools/xaprepare/xaprepare/ConfigAndData/Configurables.MacOS.cs#L10

I guess this doc is out of date now? https://github.com/xamarin/xamarin-android/blob/master/Documentation/workflow/HowToBumpMono.md#update-system-mono

Whoever does the integration this time around, should update it for the next person. 😄

**Comment by radical** [comment]
This should be because of:

```
"/Users/builder/azdo/_work/3/s/xamarin-android/Xamarin.Android.sln" (default target) (1) ->
"/Users/builder/azdo/_work/3/s/xamarin-android/src/Xamarin.Android.Build.Tasks/Xamarin.Android.Build.Tasks.csproj" (default target) (17) ->
(CoreCompile target) -> 
  /Users/builder/azdo/_work/3/s/xamarin-android/external/mono/sdks/out/android-sources/external/linker/src/linker/Linker/LinkContext.cs(154,10): error CS0246: The type or namespace name 'IReflectionPatternRecorder' could not be found (are you missing a using directive or an assembly reference?) [/Users/builder/azdo/_work/3/s/xamarin-android/src/Xamarin.Android.Build.Tasks/Xamarin.Android.Build.Tasks.csproj]
  /Users/builder/azdo/_work/3/s/xamarin-android/external/mono/sdks/out/android-sources/external/linker/src/linker/Linker/Tracer.cs(60,28): error CS0246: The type or namespace name 'IDependencyRecorder' could not be found (are you missing a using directive or an assembly reference?) [/Users/builder/azdo/_work/3/s/xamarin-android/src/Xamarin.Android.Build.Tasks/Xamarin.Android.Build.Tasks.csproj]
  /Users/builder/azdo/_work/3/s/xamarin-android/external/mono/sdks/out/android-sources/external/linker/src/linker/Linker/Tracer.cs(39,8): error CS0246: The type or namespace name 'IDependencyRecorder' could not be found (are you missing a using directive or an assembly reference?) [/Users/builder/azdo/_work/3/s/xamarin-android/src/Xamarin.Android.Build.Tasks/Xamarin.Android.Build.Tasks.csproj]
```

The `Xamarin.Android.Build.Tasks` project fails to build completely, thus `Xamarin.Android.CSharp.targets` never becomes available for other projects to use, causing projects like `NUnitLite` to fail.

Linker (via mono?) probably needs a bump here.

**Comment by fanyang-mono** [comment]
@monojenkins build failed

**Comment by fanyang-mono** [comment]
@monojenkins build failed

**Comment by fanyang-mono** [comment]
@monojenkins build

File: src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/FixAbstractMethodsStep.cs [hunk 1]
```diff
@@ -281,11 +281,6 @@ void AddNewExceptionMethod (TypeDefinition type, MethodDefinition method)
 			}
 		}
 
-		public virtual void LogMessage (string message, params object [] values)
```
Comment (radekdoulik) [inline src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/FixAbstractMethodsStep.cs]: This method is overridden in https://github.com/xamarin/xamarin-android/blob/master/src/Xamarin.Android.Build.Tasks/Tasks/LinkAssembliesNoShrink.cs#L108

**Comment by EgorBo** [comment]
@monojenkins build

**Comment by radical** [comment]
`Windows`:
`src\Xamarin.Android.NUnitLite\Xamarin.Android.NUnitLite.csproj` uses `Xamarin.Android.CSharp.targets` from the output directory. This project fails because the targets file doesn't exist, which is because `Xamarin.Android.Build.Tasks` project gets built later! There needs to be dependency to ensure build order here.

**Comment by radical** [comment]
Trying a fix for the windows issue.

**Comment by radical** [comment]
@rainersigwald any better solution for this - https://github.com/xamarin/xamarin-android/pull/4219/commits/ddd8e28319b32f1cea780bcac73d282b1d89ef99 ?

**Comment by rainersigwald** [comment]
@radical Sorry, missed this. The situation is that `Build.Tasks.csproj` generates a file that's imported by other projects? So it has to be completely _built_ before the other projects are _loaded?_ Then the solution dependency is probably best. That's not a great situation to be in for that reason--do the projects load in VS at all?

**Comment by radical** [comment]
> @radical Sorry, missed this. The situation is that `Build.Tasks.csproj` generates a file that's imported by other projects? So it has to be completely _built_ before the other projects are _loaded?_ Then the solution dependency is probably best. That's not a great situation to be in for that reason--do the projects load in VS at all?

Um @jonpryor ^ ?

**Comment by jonpryor** [comment]
What's "funny" is that we *used* to have `Xamarin.Android.NUnitLite.csproj` depend on `Xamarin.Android.Build.Tasks.csproj`, but that was removed in 91467bbdd695c3fb92d2aed3327a3d2c64b01c1e:

https://github.com/xamarin/xamarin-android/commit/91467bbdd695c3fb92d2aed3327a3d2c64b01c1e#diff-20995ab420e714d4a2d1360ae79fd1e6L362

This in turn raises a different question: why was that `@(ProjectReference)` conditional in the first place?  It was made conditional in 3ed220d5246f45dd29b15333e416890bda7ff68f, but I do not know *why* it was made conditional.

**Comment by jpobst** [comment]
I was able to eventually load `Xamarin.Android.sln` is VS 16.4.6.  It was slow and gave this error:

Annotation 2020-03-11 142148

This error does not appear in `master`.

Note that you do `msbuild Xamarin.Android.sln /t:Prepare` from the command line before you open it in VS.  I think that prebuilds some of the things that VS needs to have built before it can load the solution.

**Comment by jonpryor** [comment]
Squash-and-merge summary:

```
Bump to mono/mono/2020-02@d90665a4 (#4219)
```

Body:

```
Changes: https://github.com/mono/api-snapshot/compare/53a841ffd666d32cfa893bcd0a686ec69d91a8eb...5b8247e289d17cef9ff645e21d896ea75bfe6328

	$ git diff --shortstat 53a841ff...5b8247e2
	 9 files changed, 572 insertions(+), 13 deletions(-)

Changes: https://github.com/mono/corefx/compare/1cdb9c20946c74a88a7fa507215556fadc2c2b27...7c9e2158925339a5a96ff8d09d21c600a0b27e5a

	$ git diff --shortstat 1cdb9c20...7c9e2158
	 24 files changed, 2393 insertions(+), 396 deletions(-)

Changes: https://github.com/mono/cecil/compare/a6a7f5c0f48c2cbbec699726bbac0816ae097003...8021f3fbe75715a1762e725594d8c00cce3679d8

	$ git diff --shortstat a6a7f5c0...8021f3fb
	 16 files changed, 98 insertions(+), 25 deletions(-)

Changes: https://github.com/mono/linker/compare/e8d054bf3224f6cda6f5ecaf420f7c9d65ade70f...e1c7a72951c86db5c6df5ea8e1864432f7a6b12d

	$ git diff --shortstat e8d054bf...e1c7a729
	 220 files changed, 9758 insertions(+), 3165 deletions(-)

Changes: https://github.com/mono/mono/compare/2ff898859b8b75f123a5c895959a7e82dc829ea5...d90665a422e9f8d015585b3ca381d74faa033cc4

	$ git diff --shortstat 2ff89885...d90665a4
	 612 files changed, 20193 insertions(+), 10239 deletions(-)

Context: https://github.com/mono/mono/issues/9726
Context: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1048838
Context: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1050615
Context: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1069059
Context: https://github.com/mono/mono/issues/10643
Context: https://github.com/mono/mono/issues/10651
Context: https://github.com/mono/mono/issues/12022
Context: https://github.com/mono/mono/issues/12995
Context: https://github.com/mono/mono/issues/15612
Context: https://github.com/mono/mono/issues/16513
Context: https://github.com/mono/mono/issues/16588
Context: https://github.com/mono/mono/issues/16778
Context: https://github.com/mono/mono/issues/16969
Context: https://github.com/mono/mono/issues/17140
Context: https://github.com/mono/mono/issues/17601
Context: https://github.com/mono/mono/issues/17869
Context: https://github.com/mono/mono/issues/17878
Context: https://github.com/mono/mono/issues/17916
Context: https://github.com/mono/mono/issues/17926
Context: https://github.com/mono/mono/issues/17980
Context: https://github.com/mono/mono/issues/18006
Context: https://github.com/mono/mono/issues/18019
Context: https://github.com/mono/mono/issues/18020
Context: https://github.com/mono/mono/issues/18030
Context: https://github.com/mono/mono/issues/18061
Context: https://github.com/mono/mono/issues/18064
Context: https://github.com/mono/mono/issues/18106
Context: https://github.com/mono/mono/issues/18120
Context: https://github.com/mono/mono/issues/18191
Context: https://github.com/mono/mono/issues/18202
Context: https://github.com/mono/mono/issues/18213
Context: https://github.com/mono/mono/issues/18221
Context: https://github.com/mono/mono/issues/18247
Context: https://github.com/mono/mono/issues/18273
Context: https://github.com/mono/mono/issues/18276
Context: https://github.com/mono/mono/issues/18317
Context: https://github.com/mono/mono/issues/18323
Context: https://github.com/mono/mono/issues/18364
Context: https://github.com/mono/mono/issues/18370
Context: https://github.com/mono/mono/issues/18417
Context: https://github.com/mono/mono/issues/18418
Context: https://github.com/mono/mono/issues/18455
Context: https://github.com/mono/mono/issues/18506
Context: https://github.com/mono/mono/issues/18524
Context: https://github.com/mono/mono/issues/18530
Context: https://github.com/mono/mono/issues/18554
Context: https://github.com/mono/mono/issues/18572
Context: https://github.com/mono/mono/issues/18578
Context: https://github.com/mono/mono/issues/18584
Context: https://github.com/mono/mono/issues/18612
Context: https://github.com/mono/mono/issues/18614
Context: https://github.com/mono/mono/issues/18675
Context: https://github.com/mono/mono/issues/18676
Context: https://github.com/mono/mono/issues/18925
Context: https://github.com/mono/mono/issues/19009
Context: https://issuetracker.unity3d.com/issues/unity-physics-collisions-do-not-work-and-errors-are-thrown-when-entering-play-mode
Context: https://xamarin.github.io/bugzilla-archives/20/20233/bug.html
```

**Pull request merged**